Early stopping in ShapRFECV

operte opened the pull request:
There are two lines where a classifier is fitted.
This commit creates a new method that performs the fitting, which can
help to add functionality to the fitting step in the future.

The initial idea was for `_fit_clf` to be a normal method. However, in
that case it could not be called as `self._fit_clf` inside
`_get_feature_shap_values_per_fold` because it is a static method and it
does not recognize `self`.
This might not be the best approach, since, for example, when
implementing the Early Stopping version of ShapRFECV, we might have some
arguments inside `self` which we want to access.

operte reviewed (commented):
I reverted this change because I don't know how to allow early stopping during the optimization step. Early stopping requires defining an eval set and this would complicate our code.

If you agree that this is not crucial for now, I suggest just letting the user know that early stopping is not passed to the optimization step, just to the shap values estimation one. We can let users know that they can overload this step if they deem it important, in the documentation.

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -695,3 +695,157 @@ def plot(self, show=True, **figure_kwargs):
+
+        # Fit model with train folds
+        if self.early_stopping_rounds:
+            clf = clf.fit(
```
I have another idea: 
How about we package lines between 833 and 843 into a separate function `_fit_model`. 
This way we don't need to copy paste the code of `_get_feature_shap_values_per_fold` and only make sure we pass kwargs to the `_fit_model` correctly. What do you think?

The main reason for doing this is that we also need to pass early stopping to `fit` of ShapRFECV in lines 502 and 506.

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -500,7 +521,7 @@ def fit(self, X, y, columns_to_keep=None, column_names=None, **shap_kwargs):
             # Optimize parameters
             if self.search_clf:
-                current_search_clf = clone(self.clf).fit(current_X, self.y)
+                current_search_clf = self._fit_clf(clone(self.clf), current_X, self.y)
```
Hmm indeed that is the tough part.

You could also consider having hold out of 10/20%. There is a parameter in xgb and lgbm that modify the size of the early stopping hold out. This way we can try to pass it to the gridsearch. However, I also think that this is really tricky, and I would consider even completely removing option of optimization for `EarlyStoppingShapRFECV`. Early stopping on its own is a regularization technique and you could argue you don't need to optimize so much there. At least if you want to have it run very quickly and not overfit. Speed and regularization would be the main benefits of using that, and if the user prefers steady optimization he can choose ShapRFECV.

operte commented:
- made `_get_feature_shap_values_per_fold` a normal method both for `ShapRFECV` and `EarlyStoppingShapRFECV`. 
- `self.search_clf` is always `False` for  `EarlyStoppingShapRFECV`. Added a warning on the `init` about this.

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -695,3 +681,158 @@ def plot(self, show=True, **figure_kwargs):
+        # Fit model with train folds
+        clf = clf.fit(
+            X_train,
+            y_train,
```
I think you forgot the early stopping rounds

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -695,3 +681,158 @@ def plot(self, show=True, **figure_kwargs):
+                    "The hyperparameter tuning step will be skipped. "
+                    "You can use ShapRFECV for that purpose."
+                )
+        self.search_clf = False
```
Let's not set it to always false.

Maybe we can allow user to make hyperparam optimization before the CV SHap estimation,  but just warn him about the fact that it is not going to be used in calculating the performance and CV SHAP values.

The idea is:
You can optimize height of the tree without early stopping, etc,  but in the final thing it is going to run the optimized parameters already + early stopping.

And also overwriting it is not fully safe, because if someone passes RandomizedSearchCV, the clf is not going to extracted from it this way and there will be a very weird error.

So my proposal:
warning that: optimization is done only before without early stopping, and then in CV we use only early stopping. and not overwriting self.search_clf

What do you think?

operte commented:
- updated the warning for `BaseSearchCV`
- fixed a bug in `ShapREFCV._get_feature_shap_values_per_fold`
- updated tests for `test_get_feature_shap_values_per_fold` since this is not a static method anymore

I'll write some tests for `EarlyStoppingShapRFECV`

operte commented:
I'm trying to write tests for `EarlyStoppingShapRFECV` but I can't use the default `X,y` datasets and `complex_lightgbm`. I get an `AttributeError: 'TreeEnsemble' object has no attribute 'values'`. 

I tried a bunch of things to fix this. It seems related to the small size of `X,y`. 
If I try testing with `complex_data` and `cv>=10` this error goes away, but it comes back for `cv<10`.

@Matgrb did you ever see this behaviour before?

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -695,3 +680,156 @@ def plot(self, show=True, **figure_kwargs):
+    def __init__(
+        self,
+        clf,
+        early_stopping_rounds=30,
```
What is the reasonable default for this value. What is it set as in LGBm and XGB?
Let's use the ones they use as default.

Matgrb reviewed (changes requested):
Great work so far! 

I left some comments,  Please also make a short section in tutorial notebook for ShapRFECV and indicate there at the beginning that:
- Overall ShapRFECV is the one that has the broadest use
- EarlyStoppingRFECV is useful if you want to use LGBM and XGB, and allows for very fast computation for large datasets.
Also at the top of docs/api/feature_elimination please mention that we have 2 options for feature elimination and their applications

Overall comment for the examples we give for this use case:
- let's always make sure the number of trees in the model is high, and complexity of the trees is low. Only then you get the real benefits of early stopping.

I look forward for the next round of review.

operte commented:
I updated a few things based on your comments. Could you go over all comments and resolve the ones you think are closed? 

Still missing:
- use the scoring metric instead of passing a metric just for early stopping
- check if `clf` supports early stopping
- make `shaprfecv` show up before `earlystoppingshaprfecv` in docs
- set a proper default for `early_stopping_rounds`
- discussion at top of `api/feature_elimination`
- tutorial

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -695,3 +680,245 @@ def plot(self, show=True, **figure_kwargs):
+
+
+    # Prepare model and parameter search space
+    clf = LGBMClassifier(n_estimators=200)
```
I think this one is not gonna run correctly on macos github actions. Macos has troubles with running Lightgbm. 

I propose making a separate function for testing docstrings with lightgbm in `tests/docs/test_docstring.py`, with the following fixture:
```
@pytest.mark.skipif(os.environ.get("SKIP_LIGHTGBM") == "true", reason="LightGBM tests disabled")
```
Then the lgbm tests are going to be skipped for mac only.

Possibly the same will be required in `tests/docs/test_notebooks.py`

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -695,3 +680,156 @@ def plot(self, show=True, **figure_kwargs):
+                because it by default handles the missing values and categorical variables. This parameter also supports
+                any hyperparameter search schema that is consistent with the sklearn API e.g.
+                [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html),
+                [RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)
```
Maybe just to make it a bit clearer:
If hyperparemeter optimization is passed, the model is first tuned without early stopping, and then early stopping is applied during Shap values estimation for feature elimination. We recommend simply passing the model without hyperparameter optimization, or using ShapRFECV without early stopping.

Matgrb reviewed (changes requested):
Nice to see the progress! The code already looks good, couple of small fixes and the docs.

operte commented:
- updated docs
- tests are now skipped if environment doesn't support LGBM
- raise error if early_stopping_rounds = None and refer user to ShapRFECV

Still missing:
- use the scoring metric instead of passing a metric just for early stopping
- check if clf supports early stopping
- make shaprfecv show up before earlystoppingshaprfecv in docs
- discussion at top of api/feature_elimination
- tutorial

Matgrb commented on probatus/feature_elimination/feature_elimination.py:
```diff
@@ -695,3 +680,259 @@ def plot(self, show=True, **figure_kwargs):
+    def __init__(
+        self,
+        clf,
+        early_stopping_rounds=None,
```
maybe we can set it to something that makes sense, and allows users to run it straight away without error?

How about `5`?

Matgrb commented on tests/feature_elimination/test_feature_elimination.py:
```diff
@@ -248,3 +250,106 @@ def test_complex_dataset(complex_data, complex_lightgbm):
+    """
+    Test EarlyStoppingShapRFECV with a LGBMClassifier.
+    """
+    clf = complex_lightgbm
```
I think it is better to not use this fixture, but rather initialize your own lgbm with 200 trees and max depth of 3, so you actually cause early stopping to kick in.

Matgrb reviewed (changes requested):
Looks very good, couple of small corrections and we can merge and make a release 1.8.0

Matgrb commented:
Great job @operte  ! We have a new feature in probatus.

I will make a release 1.8.0 today with the new feature
